src/static/sw.js

```diff
@@ -0,0 +1,83 @@
+// I am a service worker for the remote newtab page.
+"use strict";
+
+// NOTE: To invalidate cached resources, change the version number below.
+const VERSION = '1';
+const CACHE_NAME = 'remote-newtab-v' + VERSION;
+
+self.addEventListener('install', function (event) {
+  let resourcesToCache = ['./'];
+
+  event.waitUntil(
+    // Fetch the list of resources to cache.
+    fetch('./mainSiteURLs.json')
+      .then(function(response){
+        return response.json();
+      }).then(function(data) {
+        resourcesToCache = resourcesToCache.concat(data);
+        caches.open(CACHE_NAME).then(function (cache) {
+          return cache.addAll(resourcesToCache);
+        })
+      })
+  );
+});
+
+function networkFirst(evt) {
+  // Current browser implementations require waitUntil() to be called
+  // synchronously during the event dispatch.  In the future we should
+  // be able to move the evt.waitUntil() down to directly where we
+  // need it in the respondWith() async handling.
+  var waitUntilResolve;
+  evt.waitUntil(new Promise(function(resolve) {
+    waitUntilResolve = resolve;
+  }));
+
+  var cache;
+  evt.respondWith(caches.open(CACHE_NAME).then(function(c) {
```

Comment:
If I were marcos I'd be all like, "use a generator!" but since I'm not marcos I won't :stuck_out_tongue_winking_eye: